File: flex/interactive/sdk/java/src/main/java/com/alibaba/graphscope/interactive/client/common/Status.java
/*
 * Copyright 2022 Alibaba Group Holding Limited.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  	http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.alibaba.graphscope.interactive.client.common;

import com.alibaba.graphscope.interactive.ApiException;
import com.alibaba.graphscope.interactive.ApiResponse;
import com.alibaba.graphscope.interactive.proto.Code;

/**
 * Mapping http status code to our status code, along with a message
 */
public class Status {

    private final Code code;
    private final String message;

    public Status() {
        this.code = Code.OK;
        this.message = "";
    }

    public boolean IsOk() {
        return this.code == Code.OK;
    }

    public String getMessage() {
        return message;
    }

    public Code getCode() {
        return this.code;
    }

    public Status(Code code, String message) {
        this.code = code;
        this.message = message;
    }

    public static Status ok(String message) {
        return new Status(Code.OK, message);
    }

    public static Status badRequest(String message) {
        return new Status(Code.BAD_REQUEST, message);
    }

    public static Status serverInternalError(String message) {
        return new Status(Code.INTERNAL_ERROR, message);
    }

    public static Status fromException(ApiException exception) {
        // mapping exception's http code to our status code
        switch (exception.getCode()) {
            case 400:
                return new Status(Code.BAD_REQUEST, exception.getMessage());
            case 403:
                return new Status(Code.PERMISSION_DENIED, exception.getMessage());
            case 404:
                return new Status(Code.NOT_FOUND, exception.getMessage());
            case 503:
                return new Status(Code.SERVICE_UNAVAILABLE, exception.getMessage());
            default:
                return new Status(Code.INTERNAL_ERROR, exception.getMessage());
        }
    }

    public static <T> Status fromResponse(ApiResponse<T> response) {
        // mapping response's http code to our status code
        switch (response.getStatusCode()) {
            case 200:
                return new Status(Code.OK, "OK");
            case 400:
                return new Status(Code.BAD_REQUEST, "Bad request");
            case 403:
                return new Status(Code.PERMISSION_DENIED, "Permission denied");
            case 404:
                return new Status(Code.NOT_FOUND, "Not found");
            case 503:
                return new Status(Code.SERVICE_UNAVAILABLE, "Service unavailable");
            default:
                return new Status(Code.INTERNAL_ERROR, "Internal error");
        }
    }

    public static Status Ok() {
        return new Status(Code.OK, "");
    }
}

File: flex/interactive/sdk/python/gs_interactive/client/status.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# Copyright 2020 Alibaba Group Holding Limited. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

from gs_interactive.api_response import ApiResponse
from gs_interactive.exceptions import ApiException
from gs_interactive.exceptions import BadRequestException
from gs_interactive.exceptions import ForbiddenException
from gs_interactive.exceptions import NotFoundException
from gs_interactive.exceptions import ServiceException
from urllib3.exceptions import MaxRetryError

from gs_interactive.client.generated.interactive_pb2 import Code as StatusCode
from gs_interactive.models.api_response_with_code import APIResponseWithCode


class Status:
    """
    This class represents the status of an operation. It contains the status code and the message.
    """

    def __init__(self, status: StatusCode, message: str):
        """
        Construct a new Status object with the specified status code and message.

        Args:
            status (StatusCode): the returnd code of the operation.
            message (str): the message returned by the operation.
        """
        self.status = status
        self.message = message

    def __str__(self):
        return f"Status: {self.status}, message: {self.message}"

    def __repr__(self):
        return f"Status: {self.status}, message: {self.message}"

    def is_ok(self) -> bool:
        """
        Whether the operation is successful.
        """

        return self.status == StatusCode.OK

    def is_error(self) -> bool:
        """
        Whether the operation is failed.
        """
        return self.status != StatusCode.OK

    def get_code(self):
        """
        Get the status code returned by the operation.
        """

        return self.status

    @property
    def get_message(self):
        """
        Get the message returned by the operation.
        """

        return self.message

    # static method create a server internal error object
    @staticmethod
    def server_internal_error(message: str):
        """
        Create a server internal error object with the specified message.
        """

        return Status(StatusCode.INTERNAL_ERROR, message)

    @staticmethod
    def from_exception(exception: ApiException):
        """
        Create a Status object from an ApiException.
        """

        # mapping from ApiException to StatusCode
        print("exception: ", exception)
        if isinstance(exception, BadRequestException):
            return Status(StatusCode.BAD_REQUEST, exception.body)
        elif isinstance(exception, ForbiddenException):
            return Status(StatusCode.PERMISSION_DENIED, exception.body)
        elif isinstance(exception, NotFoundException):
            return Status(StatusCode.NOT_FOUND, exception.body)
        elif isinstance(exception, ServiceException):
            if exception.status == 503:
                return Status(StatusCode.SERVICE_UNAVAILABLE, exception.body)
            else:
                return Status(StatusCode.INTERNAL_ERROR, exception.body)
        elif isinstance(exception, MaxRetryError):
            return Status(StatusCode.INTERNAL_ERROR, exception)
        return Status(
            StatusCode.UNKNOWN, "Unknown Error from exception " + exception.body
        )

    @staticmethod
    def from_response(response: ApiResponse):
        """
        Create a Status object from an ApiResponse.
        """

        # mapping from ApiResponse to StatusCode
        if response.status_code == 200:
            return Status(StatusCode.OK, "OK")
        else:
            # If the status_code is not 200, we expect APIReponseWithCode returned from server
            api_response_with_code = response.data
            if isinstance(api_response_with_code, APIResponseWithCode):
                return Status(
                    api_response_with_code.code, api_response_with_code.message
                )
            return Status(StatusCode.UNKNOWN, "Unknown Error")

    @staticmethod
    def ok():
        """
        Create a successful status object.
        """

        return Status(StatusCode.OK, "OK")
